File: exercises/csharp/TriviaGame/TriviaGame/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace TriviaGame
{
    public class Game
    {
        readonly List<Player> players = new List<Player>();

        private const int WINNING_TOTAL = 6;
        private const int MINIMUM_NUMBER_OF_PLAYERS = 2;
        private const int TOTAL_NUMBER_OF_QUESTIONS = 50;
        private const int TOTAL_PLACES_OF_BOARD = 12;
        private int currentPlayer;
        private bool isGettingOutOfPenaltyBox;

        readonly Deck popQuestions = new Deck("Pop");
        readonly Deck scienceQuestions = new Deck("Science");
        readonly Deck sportsQuestions = new Deck("Sports");
        readonly Deck rockQuestions = new Deck("Rock");

        public Game()
        {
            for (int questionNumber = 0; questionNumber < TOTAL_NUMBER_OF_QUESTIONS; questionNumber++)
            {
                popQuestions.AddQuestion(questionNumber);
                scienceQuestions.AddQuestion(questionNumber);
                sportsQuestions.AddQuestion(questionNumber);
                rockQuestions.AddQuestion(questionNumber);
            }
        }

        public bool IsPlayable()
        {
            return (NumberOfPlayers() >= MINIMUM_NUMBER_OF_PLAYERS);
        }

        public bool Add(string playerName)
        {
            players.Add(new Player(playerName));

            Console.WriteLine(playerName + " was added");
            Console.WriteLine("They are player number " + players.Count);
            return true;
        }

        public int NumberOfPlayers()
        {
            return players.Count;
        }

        public void Roll(int roll)
        {
            Console.WriteLine(players[currentPlayer].Name + " is the current player");
            Console.WriteLine("They have rolled a " + roll);

            bool isRollOdd = roll % 2 != 0;
            bool IsInPenaltyBox = players[currentPlayer].InPenaltyBox;

            if (IsInPenaltyBox)
            {
                DecideIfPlayerGettingOutOfPenantyBox(isRollOdd);

            }

            bool isInBoxButGettingOut = IsInPenaltyBox && isGettingOutOfPenaltyBox;
            if (!IsInPenaltyBox || isInBoxButGettingOut)
            {
                MoveToNewLocation(roll);
                AskQuestion();
            }

        }

       
        public bool IsAnswerCorrect()
        {
            if (players[currentPlayer].InPenaltyBox)
            {
                return DecideIfPlayerWon();
            }
            else
            {
                Console.WriteLine("Answer was corrent!!!!");
                return IsWinner();
            }
        }

        
        public bool IsAnswerWrong()
        {
            Console.WriteLine("Question was incorrectly answered");
            Console.WriteLine(players[currentPlayer].Name + " was sent to the penalty box");
            players[currentPlayer].InPenaltyBox = true;
            CurrentPlayerCount();
            return true;
        }

        private void DecideIfPlayerGettingOutOfPenantyBox(bool isRollOdd)
        {
            if (isRollOdd)
            {
                isGettingOutOfPenaltyBox = true;

                Console.WriteLine(players[currentPlayer].Name + " is getting out of the penalty box");
            }
            else
            {
                Console.WriteLine(players[currentPlayer].Name + " is not getting out of the penalty box");
                isGettingOutOfPenaltyBox = false;
            }
        }

        private void MoveToNewLocation(int roll)
        {
            players[currentPlayer].Place = players[currentPlayer].Place + roll;
            if (players[currentPlayer].Place > TOTAL_PLACES_OF_BOARD - 1) players[currentPlayer].Place = players[currentPlayer].Place - TOTAL_PLACES_OF_BOARD;

            Console.WriteLine(players[currentPlayer].Name
                    + "'s new location is "
                    + players[currentPlayer].Place);
            Console.WriteLine("The category is " + CurrentCategory());
        }
        private bool DecideIfPlayerWon()
        {
            if (isGettingOutOfPenaltyBox)
            {
                Console.WriteLine("Answer was correct!!!!");
                return IsWinner();
            }
            else
            {
                CurrentPlayerCount();
                return true;
            }
        }

        private void AskQuestion()
        {
            if (CurrentCategory() == "Pop")
            {
                AskAndRemoveQuestion(popQuestions.Questions);
            }
            if (CurrentCategory() == "Science")
            {
                AskAndRemoveQuestion(scienceQuestions.Questions);
            }
            if (CurrentCategory() == "Sports")
            {
                AskAndRemoveQuestion(sportsQuestions.Questions);
            }
            if (CurrentCategory() == "Rock")
            {
                AskAndRemoveQuestion(rockQuestions.Questions);
            }
        }

        private void AskAndRemoveQuestion(LinkedList<string> questions)
        {
            Console.WriteLine(questions.First());
            questions.RemoveFirst();
        }

        private string CurrentCategory()
        {
            if (players[currentPlayer].Place == 0) return "Pop";
            if (players[currentPlayer].Place == 4) return "Pop";
            if (players[currentPlayer].Place == 8) return "Pop";
            if (players[currentPlayer].Place == 1) return "Science";
            if (players[currentPlayer].Place == 5) return "Science";
            if (players[currentPlayer].Place == 9) return "Science";
            if (players[currentPlayer].Place == 2) return "Sports";
            if (players[currentPlayer].Place == 6) return "Sports";
            if (players[currentPlayer].Place == 10) return "Sports";
            return "Rock";
        }

        private bool IsWinner()
        {
            players[currentPlayer].Purse++;
            Console.WriteLine(players[currentPlayer].Name
                    + " now has "
                    + players[currentPlayer].Purse
                    + " Gold Coins.");

            bool winner = DidPlayerWin();
            CurrentPlayerCount();

            return winner;
        }

        private void CurrentPlayerCount()
        {
            currentPlayer++;
            if (currentPlayer == players.Count) currentPlayer = 0;
        }

        private bool DidPlayerWin()
        {
            return !(players[currentPlayer].Purse == WINNING_TOTAL);
        }
    }

}

File: exercises/nodejs/src/game.js
import generator from 'random-seed'
import { Player } from './player'

var Game = function () {
  const MAX_PLAYERS = 6
  const winningGoldCoins = 6
  var players = []
  var board = new Array(MAX_PLAYERS)
  var playersInPenaltyBox = new Array(MAX_PLAYERS)

  var popQuestions = []

  var scienceQuestions = []
  var sportsQuestions = []
  var rockQuestions = []

  var currentPlayer = 0
  var isPlayerInPenaltyBox = false

  var isWinner = function () {
    return players[currentPlayer].purse !== winningGoldCoins
  }

  var getCurrentCategory = function () {
    const categoryIndex = board[currentPlayer] % 4
    if (categoryIndex == 0) { return 'Pop' }
    if (categoryIndex  == 1) { return 'Science' }
    if (categoryIndex == 2) { return 'Sports' }
    if (categoryIndex == 3) { return 'Rock' }
  }

  this.createRockQuestion = function (index) {
    return 'Rock Question ' + index
  }

  for (var index = 0; index < 50; index++) {
    popQuestions.push('Pop Question ' + index)
    scienceQuestions.push('Science Question ' + index)
    sportsQuestions.push('Sports Question ' + index)
    rockQuestions.push(this.createRockQuestion(index))
  }

  this.isPlayable = function (howManyPlayers) {
    return howManyPlayers >= 2
  }

  this.add = function (playerName) {
    players.push(new Player(playerName))
    const player = this.playerCount() -1
    this.initializePlayer(player)
    this.logNewPlayer(playerName)
    return true
  }
  this.initializePlayer = function (player) {
    board[player] = 0
    playersInPenaltyBox[player] = false
  }

  this.logNewPlayer = function (playerName) {
    console.log(playerName + ' was added')
    console.log('They are player number ' + players.length)
  }

  this.playerCount = function () {
    return players.length
  }
  const displayQuestion = (questions) => {
    console.log(questions.shift())
  }

  this.askQuestion = function () {
  const currentCategory = getCurrentCategory()
    if (currentCategory == 'Pop') { displayQuestion(popQuestions) }
    if (currentCategory == 'Science') { displayQuestion(scienceQuestions) }
    if (currentCategory == 'Sports') { displayQuestion(sportsQuestions) }
    if (currentCategory == 'Rock') { displayQuestion(rockQuestions) }
  }

  this.logPlayerRoll = function (roll) {
    console.log(players[currentPlayer].name + ' is the current player')
    console.log('They have rolled a ' + roll)
  }

  this.movePlayer = function (roll) {
    board[currentPlayer] += roll
    if (board[currentPlayer] > 11) {
      board[currentPlayer] -= 12
    }
    console.log(players[currentPlayer].name + "'s new location is " + board[currentPlayer])
    console.log('The category is ' + getCurrentCategory())
  }

  this.roll = function (roll) {
    const isRollOdd = roll % 2 != 0
    this.logPlayerRoll(roll)

    if (playersInPenaltyBox[currentPlayer]) {
      if (isRollOdd) {
        isPlayerInPenaltyBox = true
        console.log(players[currentPlayer].name + ' is getting out of the penalty box')
        this.movePlayer(roll)
        this.askQuestion()
      } else {
        console.log(players[currentPlayer].name + ' is not getting out of the penalty box')
        isPlayerInPenaltyBox = false
      }
    } else {
      this.movePlayer(roll)
      this.askQuestion()
    }
  }

  this.changePlayer = function () {
    currentPlayer += 1
    if (currentPlayer == players.length) { currentPlayer = 0 }
  }

  this.addToPurse = function () {
    players[currentPlayer].purse += 1
    console.log(players[currentPlayer].name + ' now has ' +
        players[currentPlayer].purse + ' Gold Coins.')
  }

  this.correctAnswer = function () {
    if (playersInPenaltyBox[currentPlayer]) {
      if (isPlayerInPenaltyBox) {
        console.log('Answer was correct!!!!')
        this.addToPurse()
        var winner = isWinner()
        this.changePlayer()
        return winner
      } else {
        this.changePlayer()
        return true
      }
    } else {
      console.log('Answer was corrent!!!!')
      this.addToPurse()
      var winner = isWinner()
      this.changePlayer()
      return winner
    }
  }

  this.wrongAnswer = function () {
    console.log('Question was incorrectly answered')
    console.log(players[currentPlayer].name + ' was sent to the penalty box')
    playersInPenaltyBox[currentPlayer] = true

    this.changePlayer()
    return true
  }
}

const runGame = (i) => {
  var isWinner = false

  var game = new Game()

  game.add('Chet')
  game.add('Pat')
  game.add('Sue')

  const random = generator.create(i)

  do {
    game.roll(random.range(5) + 1)

    if (random.range(9) == 7) {
      isWinner = game.wrongAnswer()
    } else {
      isWinner = game.correctAnswer()
    }
  } while (isWinner)
}

export default runGame
